File: src/ads1292_studio/live_render.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from itertools import islice

import numpy as np

from ads1292_studio.display import EcgDisplaySettings, SoftwareFilterSettings
from ads1292_studio.plots import decimate_extrema_for_plot, decimate_for_plot, smooth_for_plot
from ads1292_studio.quality import SignalNoiseEstimate, estimate_realtime_snr
from ads1292_studio.signal_processing import HeartRateSummary, apply_software_filters, detect_r_peaks, heart_rate_summary


MIN_PEAK_DETECTION_SECONDS = 1.0
MIN_PEAK_DETECTION_SPAN_COUNTS = 1e-9
# Contact is good enough for live peak detection when at least this fraction of
# the visible window is fully attached (status == 0). Matches the quality gate's
# min_contact_ok_percent default of 95%.
MIN_CONTACT_OK_FRACTION = 0.95


@dataclass(frozen=True)
class LiveRenderFrame:
    source: str
    left: float
    right: float
    visible_x: np.ndarray
    visible_ecg: np.ndarray
    visible_ecg_plot: np.ndarray
    visible_resp_plot: np.ndarray
    visible_status: np.ndarray
    peaks: tuple[int, ...]
    peaks_x: np.ndarray
    peaks_y: np.ndarray
    plot_ecg_x: np.ndarray
    plot_ecg: np.ndarray
    plot_resp_x: np.ndarray
    plot_resp: np.ndarray
    plot_status_x: np.ndarray
    plot_status: np.ndarray
    heart_rate: HeartRateSummary
    snr: SignalNoiseEstimate | None = None


def display_signal_values(
    values: np.ndarray,
    *,
    filter_enabled: bool,
    filter_settings: SoftwareFilterSettings | None = None,
    invert: bool = False,
    gain: float = 1.0,
    sample_rate_hz: float = 500.0,
) -> np.ndarray:
    settings = filter_settings or SoftwareFilterSettings(bandpass_enabled=filter_enabled)
    display = apply_software_filters(values, sample_rate_hz, settings)
    display = np.asarray(display, dtype=float)
    scale = -float(gain) if invert else float(gain)
    if scale == 1.0:
        return display
    return display * scale


def deque_tail_array(values: deque[float] | deque[int], count: int, *, dtype: object = float) -> np.ndarray:
    visible_count = max(0, min(len(values), int(count)))
    start = len(values) - visible_count
    return np.fromiter(islice(values, start, None), dtype=dtype, count=visible_count)


def build_live_render_frame(
    *,
    indices: deque[int],
    ch1: deque[float],
    ch2: deque[float],
    status: deque[int],
    display_settings: EcgDisplaySettings,
    filter_settings: SoftwareFilterSettings,
    source: str,
    sample_rate_hz: float,
    smoothing_window: int,
    max_render_points: int,
    ecg_inverted: bool,
) -> LiveRenderFrame | None:
    if not np.isfinite(sample_rate_hz) or sample_rate_hz <= 0:
        return None  # a bad/uninitialized rate would poison the x-axis with inf/nan
    visible_count = min(
        len(indices),
        len(ch1),
        len(ch2),
        len(status),
        int(display_settings.time_window_seconds * sample_rate_hz) + 2,
    )
    if visible_count <= 0:
        return None

    visible_x = deque_tail_array(indices, visible_count, dtype=float) / sample_rate_hz
    left = max(0.0, float(visible_x[-1]) - display_settings.time_window_seconds)
    right = max(display_settings.time_window_seconds, float(visible_x[-1]))
    ecg_raw = deque_tail_array(ch2, visible_count, dtype=float)
    resp_raw = deque_tail_array(ch1, visible_count, dtype=float)
    visible_ecg = display_signal_values(
        ecg_raw,
        filter_enabled=bool(filter_settings.bandpass_enabled),
        filter_settings=filter_settings,
        invert=ecg_inverted,
        gain=display_settings.gain,
        sample_rate_hz=sample_rate_hz,
    )
    visible_resp = display_signal_values(
        resp_raw,
        filter_enabled=bool(filter_settings.bandpass_enabled),
        filter_settings=filter_settings,
        sample_rate_hz=sample_rate_hz,
    )
    visible_ecg_plot = smooth_for_plot(visible_ecg, window=smoothing_window)
    visible_resp_plot = smooth_for_plot(visible_resp, window=smoothing_window)
    visible_status = deque_tail_array(status, visible_count, dtype=float)
    # Contact is OK only when most of the window is attached, not when a single
    # sample happens to read 0 (np.any would let one good sample mask total
    # lead-off and run R-peak detection on a disconnected trace).
    contact_ok = bool(visible_status.size and np.mean(visible_status == 0.0) >= MIN_CONTACT_OK_FRACTION)
    ecg_has_signal = bool(np.ptp(visible_ecg) > MIN_PEAK_DETECTION_SPAN_COUNTS)
    if contact_ok and ecg_has_signal and visible_count >= int(MIN_PEAK_DETECTION_SECONDS * sample_rate_hz):
        peaks = tuple(
            detect_r_peaks(
                visible_ecg,
                sample_rate_hz,
                prefiltered=bool(filter_settings.bandpass_enabled),
            )
        )
    else:
        peaks = ()
    peak_indices = list(peaks)
    peaks_x = visible_x[peak_indices] if peaks else np.array([], dtype=float)
    peaks_y = visible_ecg_plot[peak_indices] if peaks else np.array([], dtype=float)
    plot_ecg_x, plot_ecg = decimate_extrema_for_plot(visible_x, visible_ecg_plot, max_render_points)
    plot_resp_x, plot_resp = decimate_extrema_for_plot(visible_x, visible_resp_plot, max_render_points)
    plot_status_x, plot_status = decimate_for_plot(visible_x, visible_status, max_render_points)
    return LiveRenderFrame(
        source=source,
        left=left,
        right=right,
        visible_x=visible_x,
        visible_ecg=visible_ecg,
        visible_ecg_plot=visible_ecg_plot,
        visible_resp_plot=visible_resp_plot,
        visible_status=visible_status,
        peaks=peaks,
        peaks_x=peaks_x,
        peaks_y=peaks_y,
        plot_ecg_x=plot_ecg_x,
        plot_ecg=plot_ecg,
        plot_resp_x=plot_resp_x,
        plot_resp=plot_resp,
        plot_status_x=plot_status_x,
        plot_status=plot_status,
        heart_rate=heart_rate_summary(peaks, sample_rate_hz),
        snr=estimate_realtime_snr(visible_ecg, sample_rate_hz=sample_rate_hz),
    )

File: core/src/view/LiveRender.cpp
// core/src/view/LiveRender.cpp
// Portable live-view frame builder: filter → smooth → gated R-peak detection → HR → SNR → decimate.
// Faithful port of live_render.py build_live_render_frame().
// Pure C++17, no Qt, no OS. All computations in double.

#include "ads1292/view/LiveRender.h"
#include "ads1292/dsp/Display.h"
#include "ads1292/dsp/EcgReview.h"
#include "ads1292/dsp/Quality.h"
#include "ads1292/view/PlotDecimate.h"

#include <algorithm>
#include <cmath>
#include <vector>

namespace ads1292::view {

namespace {

// ─── tail helper ─────────────────────────────────────────────────────────────
// Returns the last `n` elements of `v`. Assumes n <= v.size().
template <typename T>
std::vector<T> tail(const std::vector<T>& v, std::size_t n) {
    if (n == 0 || v.empty()) return {};
    std::size_t start = (v.size() >= n) ? (v.size() - n) : 0;
    return std::vector<T>(v.begin() + static_cast<std::ptrdiff_t>(start), v.end());
}

// ─── display_signal_values ───────────────────────────────────────────────────
// Applies display filter chain then gain / polarity inversion.
// Matches live_render.py display_signal_values():
//   display = apply_software_filters(raw, sr, settings)
//   scale   = invert ? -gain : gain
//   return scale == 1.0 ? display : display * scale
std::vector<double> display_signal_values(
    const std::vector<double>& raw,
    double sample_rate_hz,
    const ads1292::dsp::SoftwareFilterSettings& settings,
    bool invert,
    double gain)
{
    std::vector<double> disp = ads1292::dsp::apply_software_filters(raw, sample_rate_hz, settings);
    const double scale = invert ? -gain : gain;
    if (scale == 1.0) {
        return disp;
    }
    for (double& v : disp) {
        v *= scale;
    }
    return disp;
}

// ─── ptp (peak-to-peak) ──────────────────────────────────────────────────────
// Returns max - min. Returns 0.0 for empty vectors.
double ptp(const std::vector<double>& v) {
    if (v.empty()) return 0.0;
    double lo = v[0], hi = v[0];
    for (double x : v) {
        if (x < lo) lo = x;
        if (x > hi) hi = x;
    }
    return hi - lo;
}

} // anonymous namespace

// ─── build_live_render_frame ─────────────────────────────────────────────────
LiveRenderFrame build_live_render_frame(
    const std::vector<int>&    indices,
    const std::vector<double>& ch1,
    const std::vector<double>& ch2,
    const std::vector<int>&    status,
    const ads1292::dsp::EcgDisplaySettings&    display_settings,
    const ads1292::dsp::SoftwareFilterSettings& filter_settings,
    const std::string& source,
    double sample_rate_hz,
    int    smoothing_window,
    int    max_render_points,
    bool   ecg_inverted)
{
    LiveRenderFrame frame;
    frame.source = source;

    // Guard: invalid sample rate
    if (!std::isfinite(sample_rate_hz) || sample_rate_hz <= 0.0) {
        return frame; // valid = false
    }

    // Visible count: min of all buffer sizes and the time-window capacity
    const std::size_t window_samples =
        static_cast<std::size_t>(
            static_cast<int>(display_settings.time_window_seconds * sample_rate_hz) + 2);

    const std::size_t visible_count = std::min({
        indices.size(),
        ch1.size(),
        ch2.size(),
        status.size(),
        window_samples
    });

    if (visible_count == 0) {
        return frame; // valid = false
    }

    // ── Extract tails ─────────────────────────────────────────────────────────
    const std::vector<int>    indices_tail = tail(indices, visible_count);
    const std::vector<double> ecg_raw      = tail(ch2,     visible_count);
    const std::vector<double> resp_raw     = tail(ch1,     visible_count);
    const std::vector<int>    status_tail  = tail(status,  visible_count);

    // ── Visible x axis ────────────────────────────────────────────────────────
    frame.visible_x.resize(visible_count);
    for (std::size_t i = 0; i < visible_count; ++i) {
        frame.visible_x[i] = static_cast<double>(indices_tail[i]) / sample_rate_hz;
    }

    // ── Time window left / right ──────────────────────────────────────────────
    const double x_back = frame.visible_x.back();
    frame.left  = std::max(0.0, x_back - display_settings.time_window_seconds);
    frame.right = std::max(display_settings.time_window_seconds, x_back);

    // ── Display filter + gain/invert ──────────────────────────────────────────
    frame.visible_ecg  = display_signal_values(
        ecg_raw, sample_rate_hz, filter_settings, ecg_inverted, display_settings.gain);

    const std::vector<double> visible_resp = display_signal_values(
        resp_raw, sample_rate_hz, filter_settings, /*invert=*/false, /*gain=*/1.0);

    // ── Smoothing ─────────────────────────────────────────────────────────────
    frame.visible_ecg_plot  = smooth_for_plot(frame.visible_ecg, smoothing_window);
    frame.visible_resp_plot = smooth_for_plot(visible_resp,      smoothing_window);

    // ── Status as double ──────────────────────────────────────────────────────
    frame.visible_status.resize(visible_count);
    for (std::size_t i = 0; i < visible_count; ++i) {
        frame.visible_status[i] = static_cast<double>(status_tail[i]);
    }

    // ── Contact gate ──────────────────────────────────────────────────────────
    // contact_ok = !status.empty() && fraction(status_tail == 0) >= 0.95
    bool contact_ok = false;
    if (!status_tail.empty()) {
        std::size_t zero_count = 0;
        for (int s : status_tail) {
            if (s == 0) ++zero_count;
        }
        const double fraction = static_cast<double>(zero_count) /
                                static_cast<double>(status_tail.size());
        contact_ok = (fraction >= 0.95);
    }

    // ── Signal presence gate ──────────────────────────────────────────────────
    const bool ecg_has_signal = ptp(frame.visible_ecg) > 1e-9;

    // ── R-peak detection ──────────────────────────────────────────────────────
    // Require: contact_ok && ecg_has_signal && visible_count >= 1 second of samples
    const bool enough_duration =
        (static_cast<int>(visible_count) >= static_cast<int>(1.0 * sample_rate_hz));

    if (contact_ok && ecg_has_signal && enough_duration) {
        frame.peaks = ads1292::dsp::detect_r_peaks(
            frame.visible_ecg,
            sample_rate_hz,
            /*prefiltered=*/filter_settings.bandpass_enabled);
    }
    // else peaks stays empty

    // ── Peak markers ─────────────────────────────────────────────────────────
    frame.peaks_x.reserve(frame.peaks.size());
    frame.peaks_y.reserve(frame.peaks.size());
    for (int pk : frame.peaks) {
        frame.peaks_x.push_back(frame.visible_x[static_cast<std::size_t>(pk)]);
        frame.peaks_y.push_back(frame.visible_ecg_plot[static_cast<std::size_t>(pk)]);
    }

    // ── Decimation ────────────────────────────────────────────────────────────
    decimate_extrema_for_plot(
        frame.visible_x, frame.visible_ecg_plot, max_render_points,
        frame.plot_ecg_x, frame.plot_ecg);

    decimate_extrema_for_plot(
        frame.visible_x, frame.visible_resp_plot, max_render_points,
        frame.plot_resp_x, frame.plot_resp);

    decimate_for_plot(
        frame.visible_x, frame.visible_status, max_render_points,
        frame.plot_status_x, frame.plot_status);

    // ── Analytics ─────────────────────────────────────────────────────────────
    frame.heart_rate = ads1292::dsp::heart_rate_summary(frame.peaks, sample_rate_hz);
    frame.snr        = ads1292::dsp::estimate_realtime_snr(frame.visible_ecg, sample_rate_hz);

    frame.valid = true;
    return frame;
}

} // namespace ads1292::view
